x-pack/legacy/plugins/apm/server/lib/transaction_groups/fetcher.ts

```diff
@@ -43,34 +44,59 @@ export function transactionGroupsFetcher(options: Options, setup: Setup) {
     { '@timestamp': { order: 'desc' as const } }
   ];
 
+  const bool =
+    options.type === 'top_traces'
+      ? {
+          must_not: {
+            exists: {
+              field: PARENT_ID
+            }
+          }
+        }
+      : {};
+
   const params = mergeProjection(projection, {
     body: {
       size: 0,
-      query: {
-        bool: {
-          // prefer sampled transactions
-          should: [{ term: { [TRANSACTION_SAMPLED]: true } }]
-        }
-      },
       aggs: {
         transactions: {
           terms: {
             ...projection.body.aggs.transactions.terms,
-            order: { sum: 'desc' as const },
-            size: config.get<number>('xpack.apm.ui.transactionGroupBucketSize')
+            size: config.get<number>('xpack.apm.ui.transactionGroupBucketSize'),
+            order: {
+              'filtered_transactions>sum': 'desc'
+            }
           },
           aggs: {
-            sample: {
-              top_hits: {
-                size: 1,
-                sort
+            select: {
+              bucket_selector: {
+                buckets_path: {
+                  filtered_transactions_count: 'filtered_transactions._count'
+                },
+                script: 'params.filtered_transactions_count > 0'
```

Comment:
I'm definitely not an expert on bucket selector aggs so take my questions with a grain of salt.
Why is this filter necessary? I'd expect the bucket to be left out if the count is zero anyway. At least if the bool filter is lifted to the top of the query (but you probably have a reason for doing it like this)

For me to understand this better : shouldn't "select" be called something like "transactions_filter" or perhaps even more verbose: "empty_transactions_filter".

Reply:
This filters out traces that do not have a root transaction available in the data set. However, that's actually a bug, and one that happens because I forgot the query should explicitly include all root transactions in a separate `should` clause. I'll fix that, and add a (lengthy) comment to explain what exactly the query does.

Reply:
I will also look into creating a new projection for traces, rather than using the transaction groups filter. The difference will be much bigger, and I've got a hunch it's more clear to have a more explicit separation than what we have today.